### PR TITLE
Hackathon July 2023

### PR DESCRIPTION
**Pull request by mpvginde**
This pull request brings all developments made during the July 2023 pySTEPS-hackathon at the RMIB.  
Many thanks to all contributors: @ladc, @RubenImhoff, Simon De Kock and Ricardo  Reinoso-Rondinel.

The work was focussed on the STEPS blending routine.

**NEW FEATURES**
1. Scientificaly correct handling of following cases when using STEPS blending (linked with #309)
- No rain in radar fields | No rain in NWP fields
- No rain in rader fiels | Rain in NWP fields

**REFACTORING** (all linked with #332)
1. Initialisation of the noise cascade with random noise (with correct spatial correlation) to have correct power 
2. Renormalize radar extrapolation cascade after AR(2)-evolution to not lose power
3. STEPS Blending extrapolation component:
- Precipiation and noise cascade are recomposed before advection
- Both fields are advected
- They are decomposed again for blending with NWP cascade
This removes the loss of power in the smallest scales due to the interpolation needed for the semilagrangian advection.  
In addition computational cost is reduced because the number of extrapolation is greatly reduced.

**BUGFIXES**
- Decoding issue for loading older (RMI) ODIM HDF5-files
- Avoid accidentally masking precip arrays by passing a copy to spatial_correlation

### CODE REVIEW

File: pysteps/blending/steps.py [hunk 1]
```diff
@@ -560,336 +570,659 @@ def forecast(
...
 
-    # 4. Estimate AR parameters for the radar rainfall field
-    PHI = _estimate_ar_parameters_radar(
-        precip_cascade, ar_order, n_cascade_levels, MASK_thr
-    )
+    # 2.3 Check for zero input fields in the radar and NWP data.
+    zero_precip_radar = blending.utils.check_norain(precip, precip_thr, norain_thr)
+    # The norain fraction threshold used for nwp is the default value of 0.0,
+    # since nwp does not suffer from clutter.
+    zero_model_fields = blending.utils.check_norain(precip_models_pm, precip_thr)
```
Comment (RubenImhoff) [inline pysteps/blending/steps.py]: I have noticed that there can be some artifacts when there is very little rainfall, so I would suggest to use the same norain_thr for the NWP model fields.

**Review by RubenImhoff — COMMENTED** [review]
Great to see the PR, @mpvginde! :)

**Comment by dnerini** [comment]
some of the tests were failing for the same reason as in #352 , @mpvginde can you please merge master into this branch to see if that solves the issue?

there are however many other failing tests, are you looking into these?

File: pysteps/blending/steps.py [hunk 2]
```diff
@@ -560,336 +570,659 @@ def forecast(
...
+            n_cascade_levels,
+            bp_filter,
+            decompositor,
+            fft,
+            noise_method,
+            noise_kwargs,
+            noise_stddev_adj,
+            measure_time,
+            num_workers,
+            seed,
```
Comment (dnerini) [inline pysteps/blending/steps.py:681]: @mpvginde I added the missing `seed` parameter here, do we also need to set the seed with numpy before the call to `_init_noise` as done elsewhere in this file?

**Comment by RubenImhoff** [comment]
I made some changes which should (hopefully) fix the issues that came from the tests. @ladc, which tests did you still want to add? We might be good to go after that.

**Comment by RubenImhoff** [comment]
The tests on Ubuntu fail due to a problem with uploading the Codecov report - not sure what is causing it, but it seems unrelated to the PR.

**Comment by dnerini** [comment]
yes I agree, we can ignore the failing test in the context of this PR. If @ladc has no further comments or changes to add, I think we can merge it!

**Review by dnerini — APPROVED** [review]

File: pysteps/blending/steps.py (hunk 1 above)
Comment (dnerini) [reply pysteps/blending/steps.py]: was this resolved in the meantime?

Comment (RubenImhoff) [reply pysteps/blending/steps.py]: Yes, I added it in my commit.

**Review by RubenImhoff — APPROVED** [review]

**Comment by mpvginde** [comment]
Thanks everyone for helping with this PR!

**Pull request merged**